cli/internal/cmd/iamupgradeapply_test.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
+	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/logger"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIamUpgradeApply(t *testing.T) {
+	setupFs := func(createConfig, createTerraformVars bool) file.Handler {
+		fs := afero.NewMemMapFs()
+		fh := file.NewHandler(fs)
+		if createConfig {
+			cfg := defaultConfigWithExpectedMeasurements(t, config.Default(), cloudprovider.Azure)
+			require.NoError(t, fh.WriteYAML(constants.ConfigFilename, cfg))
+		}
+		if createTerraformVars {
+			require.NoError(t, fh.Write(
+				filepath.Join(constants.TerraformIAMWorkingDir, "terraform.tfvars"),
+				[]byte(
+					"region = \"foo\"\n"+
+						"resource_group_name = \"bar\"\n"+
+						"service_principal_name = \"baz\"\n",
+				),
+			))
+		}
+		return fh
+	}
+
+	testCases := map[string]struct {
+		fh            file.Handler
+		iamUpgrader   *stubIamUpgrader
+		configFetcher *stubConfigFetcher
+		yesFlag       bool
+		input         string
+		wantErr       bool
+	}{
+		"success": {
+			fh:            setupFs(true, true),
+			configFetcher: &stubConfigFetcher{},
+			iamUpgrader:   &stubIamUpgrader{},
+		},
+		"abort": {
+			fh:            setupFs(true, true),
+			iamUpgrader:   &stubIamUpgrader{},
+			configFetcher: &stubConfigFetcher{},
+			input:         "no",
+			yesFlag:       true,
+		},
+		"config missing": {
+			fh:            setupFs(false, true),
+			iamUpgrader:   &stubIamUpgrader{},
+			configFetcher: &stubConfigFetcher{},
+			yesFlag:       true,
+			wantErr:       true,
+		},
+		"iam vars missing": {
+			fh:            setupFs(true, false),
+			iamUpgrader:   &stubIamUpgrader{},
+			configFetcher: &stubConfigFetcher{},
+			yesFlag:       true,
+			wantErr:       true,
+		},
+		"plan error": {
+			fh: setupFs(true, true),
+			iamUpgrader: &stubIamUpgrader{
+				planErr: assert.AnError,
+			},
+			configFetcher: &stubConfigFetcher{},
+			yesFlag:       true,
+			wantErr:       true,
+		},
+		"apply error": {
+			fh: setupFs(true, true),
+			iamUpgrader: &stubIamUpgrader{
+				hasDiff:  true,
+				applyErr: assert.AnError,
+			},
+			configFetcher: &stubConfigFetcher{},
+			yesFlag:       true,
+			wantErr:       true,
+		},
+		"restore error": {
+			fh: setupFs(true, true),
+			iamUpgrader: &stubIamUpgrader{
+				hasDiff:     true,
+				rollbackErr: assert.AnError,
+			},
+			configFetcher: &stubConfigFetcher{},
+			input:         "no\n",
+			wantErr:       true,
+		},
+		"config fetcher err": {
+			fh: setupFs(true, true),
+			iamUpgrader: &stubIamUpgrader{
+				rollbackErr: assert.AnError,
+			},
+			configFetcher: &stubConfigFetcher{
+				fetchLatestErr: assert.AnError,
+			},
+			yesFlag: true,
+			wantErr: true,
+		},
```

Comment:
thanks for adding the tests. could we add the restore called check here too?

Reply:
s.a.